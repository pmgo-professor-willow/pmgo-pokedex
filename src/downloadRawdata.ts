// Node modules.
import { mkdir, writeFile } from 'node:fs/promises';
import fetch from 'node-fetch';

const downloadGameMaster = async () => {
  const url = 'https://raw.githubusercontent.com/PokeMiners/game_masters/master/latest/latest.json';
  const res = await fetch(url);
  const json = await res.json();

  await mkdir('./rawdata', { recursive: true });
  await writeFile('./rawdata/gameMaster.json', JSON.stringify(json, null, 2));
};

const downloadI18n = async (locale = 'ChineseTraditional') => {
  const url = `https://raw.githubusercontent.com/PokeMiners/pogo_assets/master/Texts/Latest%20APK/${locale}.txt`;
  const res = await fetch(url);
  const text = await res.text();

  await mkdir('./rawdata', { recursive: true });
  await writeFile('./rawdata/resources.txt', text);
};

export {
  downloadGameMaster,
  downloadI18n,
};
