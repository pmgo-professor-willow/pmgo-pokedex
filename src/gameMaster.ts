// Node modules.
import { readFileSync } from 'node:fs';
import appRoot from 'app-root-path';

interface Template {
  templateId: string;
  data: {
    templateId: string;
    [prop: string]: unknown;
  }
}

type GameMaster = Template[];

interface Resource {
  [id: string]: string;
}

type ResourceTarget =
  | 'pokemonName'
  | 'pokemonType'
  | 'pokemonClass'
  | 'pokemonCategory'
  | 'pokemonDescription'
  | 'pokemonForm'
  | 'moveName';

const resourceTargetFilterTable: Record<ResourceTarget, RegExp> = {
  pokemonName: /^pokemon_name_(\w+)/,
  pokemonType: /^pokemon_type_(\w+)/,
  pokemonClass: /^filter_key_(\w+)/,
  pokemonCategory: /^pokemon_category_(\w+)/,
  pokemonDescription: /^pokemon_desc_(\w+)/,
  pokemonForm: /^form_(\w+)|^(\w+)_pokedex_header/,
  moveName: /^move_name_(\d+)/,
  
} as const;

/**
 * GameMaster is Pokemon Go data.
 */
const getGameMaster = () => {
  const rawdata = readFileSync(`${appRoot.path}/rawdata/gameMaster.json`, 'utf-8');
  const gameMaster: GameMaster = JSON.parse(rawdata);
  return gameMaster;
};

/**
 * Resources is I18n of Pokemon Go data.
 */
const getResources = (target: ResourceTarget) => {
  const rawdata = readFileSync(`${appRoot.path}/rawdata/resources.txt`, 'utf-8');
  const filterPattern = resourceTargetFilterTable[target];

  return rawdata.match(/RESOURCE ID: [^\n]+\nTEXT: [^\n]+/sg)?.reduce<Resource>((prev, resource) => {
    const { 1: id, 2: text } = resource.match(/RESOURCE ID: ([^\n]+)\nTEXT: ([^\n]+)/)!;
    if (filterPattern.test(id)) {
      const { 1: key, 2: subKey } = id.match(filterPattern)!;
      const finalKey = key ?? subKey;
      prev[finalKey.trim()] = text.trim();
    }
    return prev;
  }, {})!;
};

export {
  getGameMaster,
  getResources,
};
