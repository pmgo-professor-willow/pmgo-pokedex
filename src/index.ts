// Node modules.
import _ from 'lodash';
import { findBestMatch } from 'string-similarity';

const pokemonList: PokemonName[] = require('../data/pokemon-list.json');
const formList: PokemonForm[] = require('../data/form-list.json');
const regionList: Region[] = require('../data/region-list.json');
const typeList: Type[] = require('../data/type-list.json');

type Locale = 'en-US' | 'zh-TW';

interface Pokemon {
  no: number;
  name: string;
  form: string;
}

const getPokemonByNo = (pokemonNo: number) => {
  return pokemonList.find((pokemon) => pokemon.no === pokemonNo);
};

const getFormsByNo = (pokemonNo: number) => {
  return formList.filter((pokemon) => pokemon.no === pokemonNo);
};

const getPokemonNameByNo = (pokemonNo: number, locale: Locale = 'zh-TW') => {
  const foundPokemon = pokemonList.find((pokemon) => pokemon.no === pokemonNo);
  return foundPokemon ? foundPokemon[locale] : null;
};

const getPokemonByFuzzyName = (pokemonName: string, targetLocale: Locale = 'zh-TW'): Pokemon => {
  const { bestMatchIndex } = findBestMatch(pokemonName, pokemonList.map((pokemon) => pokemon['en-US']));

  const matchedPokemon = pokemonList[bestMatchIndex];

  let form = "";
  regionList.forEach((region) => {
    region.patterns.forEach((pattern) => {
      if (new RegExp(pattern, 'i').test(pokemonName)) {
        form = region[targetLocale];
      }
    });
  });

  const pokemon: Pokemon = {
    no: matchedPokemon.no,
    name: matchedPokemon[targetLocale],
    form: form,
  };

  return pokemon;
};

const transPokemonName = (pokemonName: string, pokemonNo: number, targetLocale: Locale = 'zh-TW') => {
  const baseLocale: Locale = 'en-US';
  const pokemon = getPokemonByNo(pokemonNo);
  const forms = getFormsByNo(pokemonNo);

  let translatedName = pokemonName;

  if (pokemon) {
    // Replace base name.
    translatedName = translatedName.replace(
      new RegExp(pokemon[baseLocale], 'i'),
      pokemon[targetLocale],
    );
    // Replace form type.
    forms.forEach((form) => {
      if (form[baseLocale] && form[targetLocale]) {
        translatedName = translatedName.replace(
          new RegExp(form[baseLocale]!, 'i'),
          form[targetLocale]!,
        );
      }
    });
    // Replace regional.
    regionList.forEach((region) => {
      region.patterns.forEach((pattern) => {
        translatedName = translatedName.replace(
          new RegExp(pattern, 'i'),
          region[targetLocale],
        );
      });
    });
  }

  return translatedName;
};

const transType = (typeText: string, targetLocale: Locale = 'zh-TW') => {
  const targetType = typeText.toLowerCase();
  const foundType = typeList.find((type) => type.patterns.includes(targetType));
  return foundType ? foundType[targetLocale] : null;
};

export {
  getPokemonNameByNo,
  getPokemonByFuzzyName,
  transPokemonName,
  transType,
};
