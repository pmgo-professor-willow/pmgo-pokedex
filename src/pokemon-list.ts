// Node modules.
import _ from 'lodash';
import fetch from 'node-fetch';
import { parse } from 'node-html-parser';
import urlJoin from 'url-join';
// Local modules.
import { hostUrl } from './utils';

const getNameList = async () => {
  const nameListUrl = urlJoin(hostUrl, '/wiki/List_of_Chinese_Pok%C3%A9mon_names');
  const res = await fetch(nameListUrl);
  const xml = await res.text();

  const root = parse(xml);
  const pokemonItems = root.querySelectorAll('table.roundy tr');

  const nameList = pokemonItems
    .map((pokemonItem) => {
      const cells = pokemonItem.querySelectorAll('td');
      const { 0: no, 2: nameEn, 3: nameZhTw, 4: nameZhCn } = cells.map((cell) => cell.rawText.trim());

      const isLegalPokemon = /\d+/.test(no);

      return isLegalPokemon
        ? {
          no: parseInt(no),
          'en-US': nameEn,
          'zh-TW': nameZhTw,
          'zh-CN': nameZhCn,
        }
        : null;
    })
    .filter(Boolean);

  return nameList;
};

export {
  getNameList,
};
