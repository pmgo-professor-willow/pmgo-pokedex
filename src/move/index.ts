// Node modules.
import _ from 'lodash';
// Local modules.
import { CombatMove, Move } from 'src/models/move';
import { getGameMaster } from 'src/gameMaster';
import { translateName } from 'src/move/name';
import { translateType } from 'src/move/type';

const getMoves = async () => {
  const gameMaster = await getGameMaster();

  const moves = gameMaster.reduce<any[]>((prev, template) => {
    /**
     * @example input: 'V0013_MOVE_WRAP'
     */
    const matches = template.templateId.match(/^V(\d+)_MOVE_(\w+)$/);

    if (matches) {
      const { 1: noIndex, 2: moveName } = matches;
      const move = template.data.moveSettings as Move;

      prev.push({
        name: translateName(noIndex),
        no: parseInt(noIndex),
        uniqueId: move.movementId,
        type: move.type,
        power: move.power,
        accuracyChance: move.accuracyChance,
        staminaLossScalar: move.staminaLossScalar,
        durationMs: move.durationMs,
        energyDelta: move.energyDelta,
      });
    }

    return prev;
  }, []);

  return moves;
};

const getCombatMoves = async () => {
  const gameMaster = await getGameMaster();

  const combatMoves = gameMaster.reduce<any[]>((prev, template) => {
    /**
     * @example input: 'COMBAT_V0014_MOVE_HYPER_BEAM'
     */
    const matches = template.templateId.match(/^COMBAT_V(\d+)_MOVE_(\w+)$/);

    if (matches) {
      const { 1: noIndex, 2: combatMoveName } = matches;
      const combatMove = template.data.combatMove as CombatMove;

      prev.push({
        uniqueId: combatMove.uniqueId,
        no: parseInt(noIndex),
        name: translateName(noIndex),
        type: translateType(combatMove),
        power: combatMove.power,
        energyDelta: combatMove.energyDelta,
        durationTurns: combatMove.durationTurns,
        buffs: combatMove.buffs,
      });
    }

    return prev;
  }, []);

  return combatMoves;
};

const allMoves = await getMoves();

const allCombatMoves = await getCombatMoves();

const getMoveById = (uniqueId: string) => {
  const move = allMoves.find((move) => move.uniqueId === uniqueId) ?? null;
  return move;
};

const getCombatMoveById = (uniqueId: string) => {
  const combatMove = allCombatMoves.find((move) => move.uniqueId === uniqueId) ?? null;
  return combatMove;
};

const getMoveByIds = (moveUniqueIds: string[] = []) => {
  const baseMoves = moveUniqueIds.map(getMoveById);
  const combatMoves = moveUniqueIds.map(getCombatMoveById);

  const moves = baseMoves.map((baseMove) => {
    const sameFields = ['no', 'name', 'uniqueId', 'type'];
    const combatMove = combatMoves.find((move) => move.uniqueId === baseMove.uniqueId);

    return {
      ..._.pick(baseMove, sameFields),
      base: _.omit(baseMove, sameFields),
      combat: _.omit(combatMove, sameFields),
    };
  });

  return moves;
};

export {
  getMoveByIds,
};
