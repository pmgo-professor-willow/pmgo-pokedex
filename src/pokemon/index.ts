// Local modules.
import { PokemonRaw, Pokemon } from 'src/models/pokemon';
import { getGameMaster } from 'src/gameMaster';
import { translateName } from 'src/pokemon/name';
import { translateType } from 'src/pokemon/type';
import { translateClass } from 'src/pokemon/class';
import { translateForm } from 'src/pokemon/form';
import { translateCategory } from 'src/pokemon/category';
import { translateDescription } from 'src/pokemon/description';
import { formatEvolutions } from 'src/pokemon/evolution';
import { calculateCP } from 'src/pokemon/cp';
import { createTempEvolutionPokemonInstances } from 'src/pokemon/tempEvolution';
import { getMoveByIds } from 'src/move/index';

const getPokemons = () => {
  const gameMaster = getGameMaster();

  const pokemons = gameMaster.reduce<Pokemon[]>((prev, template) => {
    /**
     * @example input: 'V0001_POKEMON_BULBASAUR'
     */
    const matches = template.templateId.match(/^V(\d+)_POKEMON_(\w+)$/);

    if (matches) {
      const { 1: noIndex, 2: pokemonNameIndex } = matches;
      const pokemon = template.data.pokemonSettings as PokemonRaw;

      if (pokemon) {
        const pokemonInstance: Pokemon = {
          uniqueId: pokemon.form ? `${pokemon.pokemonId}_${pokemon.form}` : pokemon.pokemonId,
          pokemonId: pokemon.pokemonId,
          no: parseInt(noIndex),
          name: translateName(noIndex),
          types: translateType(pokemon),
          category: translateCategory(noIndex),
          description: translateDescription(noIndex),
          form: translateForm(pokemon),
          class: translateClass(pokemon),
          // Evolutions.
          familyId: pokemon.familyId,
          evolutions: formatEvolutions(pokemon),
          // Stats and moves.
          stats: pokemon.stats,
          quickMoves: getMoveByIds(pokemon.quickMoves),
          cinematicMoves: getMoveByIds(pokemon.cinematicMoves),
          eliteQuickMoves: getMoveByIds(pokemon.eliteQuickMove),
          eliteCinematicMoves: getMoveByIds(pokemon.eliteCinematicMove),
          // Extra.
          cpTable: {
            15: calculateCP(15.0, pokemon),
            20: calculateCP(20.0, pokemon),
            25: calculateCP(25.0, pokemon),
            30: calculateCP(30.0, pokemon),
            35: calculateCP(35.0, pokemon),
            40: calculateCP(40.0, pokemon),
            50: calculateCP(50.0, pokemon),
            51: calculateCP(51.0, pokemon),
          },
        };

        const hasSamePokemon = prev.some((pokemon) =>
          pokemon.pokemonId === pokemonInstance.pokemonId &&
          pokemon.form === pokemonInstance.form
        );

        if (!hasSamePokemon) {
          // Add pokemon.
          prev.push(pokemonInstance);

          // Add mega evolution pokemons.
          prev.push(...createTempEvolutionPokemonInstances(pokemon, pokemonInstance));
        }
      }
    }

    return prev;
  }, []);

  return pokemons;
};

export {
  getPokemons,
};
