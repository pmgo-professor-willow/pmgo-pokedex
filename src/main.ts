// Node modules.
import _ from 'lodash';
// Local modules.
import * as Downloader from 'src/downloadRawdata';
import { getGameMaster, getResources } from 'src/gameMaster';
import { getPokemons } from './pokemon/index';

async function main() {
  // await Downloader.downloadGameMaster();
  // await Downloader.downloadI18n();

  const pokemons = await getPokemons();

  // console.log(pokemons.filter((pokemon) => pokemon.name === '洛奇亞'));
  // console.log(pokemons.filter((pokemon) => pokemon.name === '飄浮泡泡'));
  // console.log(pokemons.filter((pokemon) => pokemon.name === '騎拉帝納'));
  console.log(JSON.stringify(pokemons.filter((pokemon) => pokemon.name === '爆肌蚊'), null, 2));
  // console.log(pokemons.filter((pokemon) => pokemon.name === '喵喵'));
}

main();
