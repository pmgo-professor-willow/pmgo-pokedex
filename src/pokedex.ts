// Node modules.
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import _ from 'lodash';
import { findBestMatch } from 'string-similarity';
// Local modules.
import { Pokemon, PokemonForm, PokemonName, Region, Type } from './models/pokemon';

type Locale = 'en-US' | 'zh-TW';

const readJson = (filename: string) => {
  const filepath = join(dirname(require.main?.filename!), '../data', filename);
  const json = JSON.parse(readFileSync(filepath, { encoding: 'utf-8' }))
  return json;
}

class Pokedex {
  private defaultTargetLocale: Locale;
  private pokemonNameList: PokemonName[];
  private pokemonList: Pokemon[];
  private formList: PokemonForm[];
  private regionList: Region[];
  private typeList: Type[];

  constructor(targetLocale: Locale = 'zh-TW') {
    this.defaultTargetLocale = targetLocale;
    this.pokemonNameList = readJson('pokemon-name-list.json');
    this.pokemonList = readJson('pokemon-list.json');
    this.formList = readJson('form-list.json');
    this.regionList = readJson('region-list.json');
    this.typeList = readJson('type-list.json');
  }

  /** Get pokemon's name by pokemon's number. */
  getPokemonNameByNo(pokemonNo: number, locale = this.defaultTargetLocale) {
    const foundPokemon = this.pokemonNameList.find((pokemon) => pokemon.no === pokemonNo);
    return foundPokemon ? foundPokemon[locale] : null;
  };

  /** Get pokemon info object by pokemon's fuzzy name. */
  getPokemonByFuzzyName = (pokemonName: string, targetLocale = this.defaultTargetLocale): Pokemon => {
    const { bestMatchIndex: bestNameIndex } = findBestMatch(pokemonName, this.pokemonNameList.map((pokemonName) => pokemonName['en-US']));
    const { bestMatchIndex: bestIndex } = findBestMatch(pokemonName, this.pokemonList.map((pokemon) => `${pokemon.name}-${pokemon.form}`));

    const matchedPokemonName = this.pokemonNameList[bestNameIndex];
    const matchedPokemon = this.pokemonList[bestIndex];

    let predictedForm: string | null = null;

    // Regional.
    this.regionList.forEach((region) => {
      region.patterns.forEach((pattern) => {
        if (new RegExp(pattern, 'i').test(pokemonName)) {
          predictedForm = region[targetLocale];
        }
      });
    });

    // Forms.
    const formCandidates = this.formList.filter((form) => {
      const isSameName = form.name === matchedPokemon.name;
      const isMatchedForm = new RegExp(_.escapeRegExp(form['en-US']), 'i').test(String(matchedPokemon.form)) || new RegExp(_.escapeRegExp(String(matchedPokemon.form)), 'i').test(String(form['en-US']));
      return isSameName && isMatchedForm;
    });
    if (formCandidates.length > 0) {
      const formText = pokemonName.replace(String(_.first(formCandidates)?.name), '');
      const { bestMatchIndex: bestFormIndex } = findBestMatch(formText, formCandidates.map((form) => String(form[targetLocale])));
      predictedForm = String(formCandidates[bestFormIndex][targetLocale]);
    }

    // Mega evolutions.
    if (/mega\s/i.test(pokemonName)) {
      predictedForm = pokemonName.replace(matchedPokemonName['en-US'], matchedPokemonName[targetLocale]);
    }
  
    const pokemon: Pokemon = {
      no: matchedPokemonName.no,
      name: matchedPokemonName[targetLocale],
      form: predictedForm,
      types: matchedPokemon.types.map((typeText) => this.transType(typeText, targetLocale)!),
    };
  
    return pokemon;
  }

  /** Translate pokemon's display name by pokemon's fuzzy name. */
  transPokemonName(pokemonName: string, targetLocale = this.defaultTargetLocale) {
    const pokemon = this.getPokemonByFuzzyName(pokemonName, targetLocale);
  
    const translatedName = pokemon.form
      ? `${pokemon.name} (${pokemon.form})`
      : pokemon.name;

    return translatedName;
  }

  /** Translate type by text of pokemon type. */
  transType(typeText: string, targetLocale = this.defaultTargetLocale) {
    const targetType = typeText.toLowerCase();
    const foundType = this.typeList.find((type) => type.patterns.includes(targetType));
    return foundType ? foundType[targetLocale] : null;
  }
}

export { Pokedex };
