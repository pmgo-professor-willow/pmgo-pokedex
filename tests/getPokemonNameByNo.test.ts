// Local modules.
import { Pokedex } from '../src/pokedex';

describe('General case', () => {
  test.each([
    1,
    2,
    3,
  ])('English (no: %s)', (no) => {
    const pokedex = new Pokedex('en-US');
    const actual = pokedex.getPokemonNameByNo(no);
    expect(actual).toMatchSnapshot();
  });

  test.each([
    1,
    2,
    3,
  ])('Chinese (no: %s)', (no) => {
    const pokedex = new Pokedex('zh-TW');
    const actual = pokedex.getPokemonNameByNo(no);
    expect(actual).toMatchSnapshot();
  });
});
