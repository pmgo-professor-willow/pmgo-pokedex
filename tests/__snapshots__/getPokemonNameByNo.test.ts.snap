// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General case Chinese (no: 1) 1`] = `"妙蛙種子"`;

exports[`General case Chinese (no: 2) 1`] = `"妙蛙草"`;

exports[`General case Chinese (no: 3) 1`] = `"妙蛙花"`;

exports[`General case English (no: 1) 1`] = `"Bulbasaur"`;

exports[`General case English (no: 2) 1`] = `"Ivysaur"`;

exports[`General case English (no: 3) 1`] = `"Venusaur"`;
